tracer/src/Datadog.Trace/Util/Http/QueryStringObfuscator.cs

```diff
@@ -0,0 +1,94 @@
+
+        private static readonly IDatadogLogger _log = DatadogLogging.GetLoggerFor(typeof(QueryStringObfuscator));
+        private static QueryStringObfuscator _instance;
+        private static bool _globalInstanceInitialized;
+        private static object _globalInstanceLock = new();
+        private readonly Obfuscator _obfuscator;
+
+        private QueryStringObfuscator(double timeout, string pattern = null)
+        {
+            pattern ??= Tracer.Instance.Settings.ObfuscationQueryStringRegex;
+            _obfuscator = new(TimeSpan.FromMilliseconds(timeout), pattern);
+        }
+
+        internal string Obfuscate(string queryString) => _obfuscator.Obfuscate(queryString);
+
+        /// <summary>
+        /// Gets or sets the global <see cref="QueryStringObfuscator"/> instance.
+        /// </summary>
+        public static QueryStringObfuscator Instance(double timeout, string pattern = null) => LazyInitializer.EnsureInitialized(ref _instance, ref _globalInstanceInitialized, ref _globalInstanceLock, () => new(timeout, pattern));
+
+        internal class Obfuscator
+        {
+            private const string ReplacementString = "<redacted>";
+            private readonly Regex _regex;
+            private readonly bool _disabled;
+            private readonly TimeSpan _timeout;
+
+            internal Obfuscator(TimeSpan timeout, string pattern = null)
+            {
+                _timeout = timeout;
+                if (string.IsNullOrEmpty(pattern))
+                {
+                    _disabled = true;
+                }
+                else
+                {
+                    _regex = new(pattern, RegexOptions.IgnoreCase | RegexOptions.Compiled, _timeout);
+                }
+            }
+
+            internal string Obfuscate(string queryString)
+            {
+                if (_disabled || string.IsNullOrEmpty(queryString))
+                {
+                    return queryString;
+                }
+
+                try
+                {
+                    queryString = queryString.Substring(0, Math.Min(queryString.Length, 200));
+                    return _regex.Replace(queryString, ReplacementString);
+                }
+                catch (RegexMatchTimeoutException e)
+                {
+                    Log($"The regex task timed out before {_timeout.TotalMilliseconds} ms and is canceled", e);
+                }
+
+                return string.Empty;
+            }
+
+            internal virtual void Log(string message = null, Exception exception = null)
```

Comment:
Doesn't seem necessary to have a method here as you call it just once.

Reply:
I use for the unit tests verify, to check it s indeed been called with the right param.. hence the virtual keyword too.. I couldnt find a better way to verify a log has indeed happened

Reply:
thanks, fixed in c5c4456c0eec985161bda650c99c0e468c5dc26b